Add nf-test to `STARFUSION_BUILD` and refactor module

atrigila opened the pull request:
This PR addresses:

- The generation of an nf-test for `STARFUSION_BUILD` #579 
- Since the original container was not working in the tests, I updated starfusion version to 1.14. 
- The code allowed for refactoring as the databases (`pfam`, `dfam`, `AnnotFilterRule.pm`) are downloaded automatically in the script call. 
- The `fusion_annot_lib` is now an external parameter that allows for easier integration with smaller datasets. Removed several lines which were hardcoded to download databases.

If this design is OK, then we will have to simply add the desired `fusion_annot_lib` (e.g. path to the full `fusion_lib.Mar2021.dat.gz`) as an input channel in the `--build_references` workflow (I can do that in this same PR). 

## Summary
Closes #579
Closes #549 

<!--
# nf-core/rnafusion pull request

Many thanks for contributing to nf-core/rnafusion!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/rnafusion/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/rnafusion/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/rnafusion _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
<!-- - [ ] Ensure the test suite passes (`nf-test test main.nf.test -profile test,docker`). -->
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

atrigila commented:
We can wait until https://github.com/nf-core/rnafusion/pull/505 is merged to add these changes, but I'll be happy to have your feedback in the meanwhile @rannick.

rannick commented on modules/local/starfusion/build/main.nf:
```diff
@@ -1,37 +1,32 @@
 
     output:
-    path "*"  , emit: reference
+    path "ctat_genome_lib_build_dir"  , emit: reference
```
Make corresponding changes on the publishing directory in modules.config so the reference folder structure remains the same

rannick commented on modules/local/starfusion/build/environment.yml:
```diff
@@ -4,7 +4,8 @@ channels:
-  - bioconda::star=2.7.11b
+  - bioconda::minimap2
+  - bioconda::samtools
+  - bioconda::star
```
I would set the versions and make sure star is the latest available release

rannick commented on modules/local/starfusion/build/main.nf:
```diff
@@ -1,37 +1,32 @@
+        --dfam_db ${dfam_species} \\
+        --pfam_db current \\
+        --fusion_annot_lib $fusion_annot_lib \\
+        ${args} \\
```
Can you write in the PR how and what you have tested this with? I will run my own build but it is good to follow the process of development

atrigila replied:
I included how to run it in a comment, let me know if you can successfully run it with your build :)

rannick commented on modules/local/starfusion/build/main.nf:
```diff
@@ -1,37 +1,32 @@
     script:
-    def binPath = (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1)  ? "prep_genome_lib.pl" : "/usr/local/src/STAR-Fusion/ctat-genome-lib-builder/prep_genome_lib.pl"
+    if (dfam_species != "human" && dfam_species != "mouse") {
+        error "Invalid species for --dfam_db. Only 'human' or 'mouse' are accepted. Provided: ${dfam_species}"
```
As a principle I think rnafusion should supports human only, but should be flexible enough to allow tweaking for other species for people to do at their own risk, so I would remove that check

atrigila replied:
This one is due to the tool. While it can accept files as input, when doing the automatic download it can only accept human or mouse: 

https://github.com/TrinityCTAT/ctat-genome-lib-builder/blob/1cf500c07a67c0240228b8f9c245c0db3e507e4f/prep_genome_lib.pl#L38

atrigila commented:
Thank you so much for your feedback @rannick! I updated the module with your suggestions and here is some extra information:
- You can try testing it with: `nextflow run . -profile test_build,docker --fusion_annot_lib <YOUR-LIB> --starfusion_build true` . I am using the small `.dat.gz` library provided in the starfusion tutorial.
- I added minimap2 because in one of the tests preparing the container, it said that it needed that library and it wasn't available.
- This is the reference folder structure now: `/outdir/references/starfusion/ctat_genome_lib_build_dir`  (which I think is the same as before).

rannick commented:
Here one issue is that the download of the necessary dependencies is not included any more outside of the test set

atrigila commented:
> Here one issue is that the download of the necessary dependencies is not included any more outside of the test set

Sorry, I am not sure I follow. Which dependencies? Are you referring to the databases `pfam`, `dfam`,` AnnotFilterRule.pm`? The script call downloads those same files automatically. As I understand, there is no need to pre-download them as the behaviour is the same.

atrigila commented on modules/local/starfusion/build/environment.yml:
```diff
@@ -4,7 +4,8 @@ channels:
-  - bioconda::trinity=2.15.2
-  - bioconda::samtools=1.21
-  - bioconda::star=2.7.11b
+  - bioconda::minimap2=2.28
```
I will re-test this just in case and adjust the environment if necessary.

rannick replied:
I just did and minimap2 is necessary

atrigila replied:
Yes, otherwise it fails without minimap2:

![image](https://github.com/user-attachments/assets/67924de1-e583-41b1-a630-a6e27d817a47)

```
    - *** ERROR, cannot locate required tool in PATH setting: minimap2 ***
```

rannick commented:
Please do not merge as if will break the logic of self contained references in the pipeline, we need to provide an alternative

grst commented on nextflow.config:
```diff
@@ -30,6 +30,7 @@ params {
     read_length                = 100
     starfusion_build           = true
     genomes                    = [:]
+    fusion_annot_lib           = null       // path to  dat.gz CTAT genome lib (e.g. https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/GRCh38_gencode_v44_CTAT_lib_Oct292023.plug-n-play.tar.gz)
```
Should `https://data.broadinstitute.org/Trinity/CTAT_RESOURCE_LIB/GRCh38_gencode_v44_CTAT_lib_Oct292023.plug-n-play.tar.gz` just become the default? Then it should still work out-of-the box @rannick 

The only difference is that now nextflow takes care of downloading the reference instead of the script calling `wget` internally.

rannick replied:
we want to update to gencode v46

rannick replied:
And do we want a 80GB file there?

grst replied:
that's fair, but would it address your comment above? 

> as if will break the logic of self contained references in the pipeline, we need to provide an alternative

rannick replied:
I was also thinking about a plug-and-play option, but if we provide a s3 option, let's have the latest one there

atrigila replied:
We can have a placeholder here until we have the latest s3 reference. For example, we could add the same original reference that was in the module (so no change in behaviour): `https://github.com/FusionAnnotator/CTAT_HumanFusionLib/releases/download/v0.3.0/fusion_lib.Mar2021.dat.gz`

and then open an issue to update it to the latest version when available in S3. Do you think that will make sense?

rannick replied:
Agree with the placeholder, it would address my comment about self-contained references. 
We can use the following link for the updated annotations it is the latest from CTAT as far as I know

https://sourceforge.net/projects/rnafusion/files/STAR-Fusion_gencodev44_CTAT_lib_Oct292925_resources_fusion_annot_lib.gz/download

rannick reviewed (approved):
I think we are ready to go!
